sdk/eventhubs/azure_messaging_eventhubs_checkpointstore_blob/src/checkpoint_store.rs

```diff
@@ -0,0 +1,324 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+//! Azure Blob Storage implementation of the Event Hubs checkpoint store.
+// cspell: ignore rfind sequencenumber ownerid
+use azure_core::{
+    http::{
+        headers::{ETAG, LAST_MODIFIED},
+        Etag, NoFormat, RequestContent, StatusCode,
+    },
+    time::parse_rfc7231,
+    Bytes, Result,
+};
+use azure_messaging_eventhubs::{
+    models::{Checkpoint, Ownership},
+    CheckpointStore,
+};
+use azure_storage_blob::{
+    models::{
+        BlobClientGetPropertiesResultHeaders, BlobClientSetMetadataOptions,
+        BlobContainerClientListBlobFlatSegmentOptions, BlockBlobClientUploadOptions,
+        BlockBlobClientUploadResultHeaders, ListBlobsIncludeItem,
+    },
+    BlobContainerClient,
+};
+use futures::TryStreamExt;
+use std::{collections::HashMap, fmt::Debug, sync::Arc};
+use time::OffsetDateTime;
+use tracing::{debug, warn};
+
+/// Azure Blob Storage implementation of the [`CheckpointStore`] trait.
+///
+/// This implementation stores checkpoint and ownership information in Azure Blob Storage,
+/// providing durable persistence for Event Hub event processing state.
+#[derive(Clone)]
+pub struct BlobCheckpointStore {
+    blob_container_client: Arc<BlobContainerClient>,
+}
+
+impl Debug for BlobCheckpointStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("BlobCheckpointStore").finish()
+    }
+}
+
+const OWNER_ID: &str = "ownerid";
+const OFFSET: &str = "offset";
+const SEQUENCE_NUMBER: &str = "sequencenumber";
+
+impl BlobCheckpointStore {
+    /// Creates a new blob checkpoint store.
+    ///
+    /// # Arguments
+    ///
+    /// * `blob_service_client` - The Azure Blob Storage service client
+    pub fn new(blob_container_client: BlobContainerClient) -> Arc<Self> {
+        Arc::new(Self {
+            blob_container_client: Arc::new(blob_container_client),
+        })
+    }
+
+    fn process_storage_response_metadata(
+        last_modified: Option<String>,
+        etag: Option<String>,
+    ) -> Result<(Option<OffsetDateTime>, Option<Etag>)> {
+        let lm = match last_modified {
+            Some(lm) => Some(parse_rfc7231(lm.as_str())?),
+            None => None,
+        };
+        Ok((lm, etag.map(Etag::from)))
+    }
+
+    async fn set_metadata_on_blob(
+        &self,
+        blob_name: &str,
+        metadata: Option<HashMap<String, String>>,
+        etag: Option<Etag>,
+    ) -> Result<(Option<OffsetDateTime>, Option<Etag>)> {
+        let blob_client = self
+            .blob_container_client
+            .blob_client(blob_name.to_string());
+        let mut options = BlobClientSetMetadataOptions::default();
+        if let Some(metadata) = &metadata {
+            options.metadata = Some(metadata.clone());
+        }
+        options.if_match = etag.clone().map(|e| e.to_string());
```

Comment:
Seems an unnecessary copy since you're just `to_string()`'ing it. Could probably use `etag.map(|ref e| e.to_string())`.